### PR TITLE
Add comprehensive automatic test

### PR DESCRIPTION
**Pull request by jameshalgren**
**TL; DR** _Added a simple, comprehensive unit test; made the t-route model scriptable_

This test is styled as a unit test, though it is more of an integration test -- if this test fails, it indicates a change more than a clear issue with any particular library. 

Perhaps of greater importance, to facilitate insertion into the test, the main function was subdivided, adding two new functions -- `_run_everything_v02` and `_handle_output_v02`. With these functions in place, the full return set from our model execution can be obtained in a deck of results organized as follows:
``` py
[list of results from each tailwater in the network
    (tuple [list of ids], [2-d array of results with dimensions: id; flow, velocity, depth for each timestep] [optionally 2-d array of results with dimensions: id; Cn, Ck, X from MC calculation for each timestep])
]
```
Converting these into a pandas dataframe, for instance, may be accomplished with the following: 
``` py

def convert_results_to_df(results, return_courant = False)
    courant = pd.Dataframe() 
    qvd_columns = pd.MultiIndex.from_product(
        [range(nts), ["q", "v", "d"]]
    ).to_flat_index()

    flowveldepth = pd.concat(
        [pd.DataFrame(r[1], index=r[0], columns=qvd_columns) for r in results],
        copy=False,
    )

    if run_parameters.get("return_courant", False):
        courant_columns = pd.MultiIndex.from_product(
            [range(nts), ["cn", "ck", "X"]]
        ).to_flat_index()
        courant = pd.concat(
            [
                pd.DataFrame(r[2], index=r[0], columns=courant_columns)
                for r in results
            ],
            copy=False,
        )
    return flowveldepth, courant
```

So... with those things in place, any script or notebook can now call the v02 execution (a v03 PR is in progress) of the t-route model and then handle the results directly, something like the following (which is exactly what is done in the new test...): 

``` py
from nwm_routing.__main__ import _run_everything_v02
from troute.nhd_io import read_custom_input
import numpy as np


def run_troute_from_script(custom_input_file):
    (
        supernetwork_parameters,
        waterbody_parameters,
        forcing_parameters,
        restart_parameters,
        output_parameters,
        run_parameters,
        parity_parameters,
        data_assimilation_parameters,
        diffusive_parameters,
        coastal_parameters,
    ) = read_custom_input(custom_input_file)

    # Modify any parameters from a template file inline here
    run_parameters["nts"] = 12

    # Then run the model and get the results
    results = _run_everything_v02(
            supernetwork_parameters,
            waterbody_parameters,
            forcing_parameters,
            restart_parameters,
            output_parameters,
            run_parameters,
            parity_parameters,
            data_assimilation_parameters,
            diffusive_parameters,
            coastal_parameters,
    )
    return results

custom_input_file = "../../test/input/yaml/CustomInput.yaml"
run_troute_from_script(custom_input_file)
fvd, courant = convert_results_to_df(results, True)

# And then do it again with another configuration or experiment 
# with comparing the pandas dataframes or whatever else you like!
```

### CODE REVIEW

**Comment by jameshalgren** [comment]
ping @BrianAvant-NOAA

**Review by awlostowski-noaa — APPROVED** [review]
This is a great new capability that greatly facilitates output comparison. Can you please give a short demo on this new feature to the Inland Team later this week?

**Comment by jameshalgren** [comment]
> This is a great new capability that greatly facilitates output comparison. Can you please give a short demo on this new feature to the Inland Team later this week?

Will do.

**Comment by BrianAvant-NOAA** [comment]
@jameshalgren can you send me an invite to the demo?
